src/kernel/capinv.c

```diff
@@ -438,21 +438,16 @@ cap_move(struct captbl *t, capid_t cap_to, capid_t capin_to,
 }
 
 static int
-cap_thd_switch(struct pt_regs *regs, struct thread *curr, struct thread  *next,
+cap_thd_switch(struct pt_regs *regs, struct thread *curr, struct thread *next,
 	       struct comp_info *ci, struct cos_cpu_local_info *cos_info)
 {
 	struct next_thdinfo *nti  = &cos_info->next_ti;
 	struct comp_info *next_ci = &(next->invstk[next->invstk_top].comp_info);
 	int               preempt = 0;
 
-	if (unlikely(curr == next)) {
-		if (!(curr->state & THD_STATE_SUSPENDED))
-			assert(!(curr->state & (THD_STATE_RCVING)) && !(curr->state & THD_STATE_PREEMPTED));
-		__userregs_set(regs, 0, __userregs_getsp(regs), __userregs_getip(regs));
-		return 0;
-	}
-
 	assert(next_ci && curr && next);
+	if (unlikely(curr == next)) return thd_switch_update(curr, regs, 1);
+
 	/* FIXME: trigger fault for the next thread, for now, return error */
```

Comment:
Is this FIXME still relevant? / Could it be easily fixed in this PR, or is fine to be left for later?